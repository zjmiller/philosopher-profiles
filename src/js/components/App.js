import React, { Component } from 'react';
import Header from './Header';
import Profiles from './Profiles';
import ProfileBig from './ProfileBig';
const sortBy = require('lodash/sortBy');
const intersection = require('lodash/intersection');

class App extends Component {
  constructor(props) {
    super(props);
    const initialState = {
      view: 'LIST',
      profile: undefined,
      filterBy: {
        gender: [],
        interests: [],
      },
      sortBy: undefined,
    };

    this.state = initialState;
  }

  enableListView = () => {
    this.setState({
      profile: undefined,
      view: 'LIST'
    });
  }

  enableGridView = () => {
    this.setState({
      profile: undefined,
      view: 'GRID'
    });
  }

  viewProfile = id => {
    this.setState({ profile: id });
  }

  leaveProfile = () => {
    this.setState({ profile: undefined });
  }

  changeFilterBy = filterBy => {
    this.setState({ filterBy });
  }

  changeSortBy = sortBy => {
    this.setState({ sortBy });
  }

  render() {
    const sortedData = do {
      if (this.state.sortBy === undefined) {
        sortBy(this.props.data.philosophers, o => o.name.last);
      }
      else if (this.state.sortBy === 'LAST_NAME') {
        sortBy(this.props.data.philosophers, o => o.name.last);
      }
      else if (this.state.sortBy === 'YEAR') {
        this.props.data.philosophers.sort((a, b) => {
          if (a.birth.era.match(/BC|BCE/) && b.birth.era.match(/BC|BCE/)) {
            if (a.birth.year > b.birth.year) return -1;
            if (a.birth.year === b.birth.year) return 0;
            return 1;
          }
          if (!a.birth.era.match(/BC|BCE/) && !b.birth.era.match(/BC|BCE/)) {
            if (a.birth.year < b.birth.year) return -1;
            if (a.birth.year === b.birth.year) return 0;
            return 1;
          }
          if (a.birth.era.match(/BC|BCE/)) return -1;
          return 0;
        });
      }
    }

    const filteredAndSortedData = sortedData
      .filter(philosopher => {
        const { gender } = this.state.filterBy;
        if (gender.length === 0) return true;
        if (gender.indexOf(philosopher.gender) > -1) return true;
        return false;
      })
      .filter(philosopher => {
        const { interests } = this.state.filterBy;
        if (interests.length === 0) return true;
        if (intersection(interests, philosopher.interests).length > 0) return true;
        return false;
      });

    return (
      <div>
        <Header
          changeFilterBy={this.changeFilterBy}
          changeSortBy={this.changeSortBy}
          enableGridView={this.enableGridView}
          enableListView={this.enableListView}
          interests={this.props.interests}
          view={this.state.view}
        />
      {
        this.state.profile !== undefined
        ?
        <ProfileBig
          leaveProfile={this.leaveProfile}
          philosopher={this.props.data.philosophers[this.state.profile]}
        />
        :
        <Profiles
          data={filteredAndSortedData}
          view={this.state.view}
          viewProfile={this.viewProfile}
        />
      }
      </div>
    );
  }
}

export default App;
