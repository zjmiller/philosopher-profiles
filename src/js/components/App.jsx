/* @flow */

import React, { Component } from 'react';
import { connect } from 'react-redux';

import Header from './Header';
import Profiles from './Profiles';
import ProfileBig from './ProfileBig';

import getFilterBy from '../selectors/getFilterBy';
import getInterests from '../selectors/getInterests';
import getPhilosophers from '../selectors/getPhilosophers';
import getPhilosopherById from '../selectors/getPhilosopherById';
import getSortBy from '../selectors/getSortBy';
import getView from '../selectors/getView';
import filterPhilosophers from '../selectors/filterPhilosophers';
import sortPhilosophers from '../selectors/sortPhilosophers';

import type { Interest, Philosopher } from '../flow-type-aliases/main';

class App extends Component {
  state: Object;

  constructor(props: Object) {
    super(props);
    const initialState = {
      profile: undefined,
    };

    this.state = initialState;
  }

  viewProfile = (id: number) => {
    this.setState({ profile: id });
  }

  leaveProfile = () => {
    this.setState({ profile: undefined });
  }

  render() {
    const interests: Interest[] = getInterests(this.props.data);
    const philosophers: Philosopher[] = getPhilosophers(this.props.data);
    const sortOpts = { sortBy: this.props.sortBy };
    const sortedPhilosophers: Philosopher[] = sortPhilosophers(philosophers, sortOpts);
    const filterOpts = {
      gender: this.props.filterBy.gender,
      interests: this.props.filterBy.interests,
    };
    const filteredAndSortedPhilosophers: Philosopher[] =
      filterPhilosophers(sortedPhilosophers, filterOpts);

    return (
      <div>
        <Header
          interests={interests}
          view={this.props.view}
        />
        {
          this.state.profile !== undefined
          ?
            <ProfileBig
              leaveProfile={this.leaveProfile}
              philosopher={getPhilosopherById(philosophers, this.state.profile)}
            />
          :
          <Profiles
            filterBy={this.props.filterBy}
            philosophers={filteredAndSortedPhilosophers}
            view={this.props.view}
            viewProfile={this.viewProfile}
          />
        }
      </div>
    );
  }
}

App.childContextTypes = {
  state: React.PropTypes.object,
};

App.propTypes = {
  data: React.PropTypes.object,
  filterBy: React.PropTypes.object,
  sortBy: React.PropTypes.string,
  view: React.PropTypes.string,
};

const mapStateToProps = state => ({
  data: state,
  filterBy: getFilterBy(state),
  sortBy: getSortBy(state),
  view: getView(state),
});

export default connect(mapStateToProps)(App);
