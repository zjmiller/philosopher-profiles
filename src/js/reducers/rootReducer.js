import { combineReducers } from 'redux';
import filterBy from './filterBy';
import sortBy from './sortBy';
import view from './view';

const interests = (state = []) => state;
const philosophers = (state = []) => state;

export default combineReducers({
  filterBy,
  interests,
  philosophers,
  sortBy,
  view,
});
