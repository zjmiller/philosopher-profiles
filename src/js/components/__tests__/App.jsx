/* eslint-disable no-param-reassign, no-undef */

import fs from 'fs';
import path from 'path';
import React from 'react';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import YAML from 'js-yaml';
import App from '../App';
import rootReducer from '../../reducers/rootReducer';

const flatten = require('lodash/flatten');
const uniq = require('lodash/uniq');

jest.mock('react-dom');

test('App renders correctly', () => {
  const responseText = fs.readFileSync(path.resolve('./dest/data.yaml'), 'utf8');
  const philosophersWithoutIds = YAML.safeLoad(responseText);

  const philosophers = philosophersWithoutIds.map((philosopher, i) =>
    Object.assign({}, philosopher, { id: i })
  );

  const arrOfArrOfInterestNames =
    philosophers.map((philosopher) => philosopher.interests);
  const arrOfInterestNames = flatten(arrOfArrOfInterestNames);
  const arrOfUniqInterestNames = uniq(arrOfInterestNames);
  const interests = arrOfUniqInterestNames.map((interestName, i) =>
    ({ id: i, name: interestName }));

  philosophers.forEach((philosopher): Philosopher => {
    philosopher.interests = philosopher.interests.map(interestName =>
       interests.find(interest => interest.name === interestName).id
    );
  });

  const preloadedState = {
    interests,
    philosophers,
    view: 'GRID',
  };

  const store = createStore(rootReducer, preloadedState);

  const tree = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
