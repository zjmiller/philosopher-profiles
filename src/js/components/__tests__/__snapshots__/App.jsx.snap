exports[`test App renders correctly 1`] = `
<div>
  <div>
    <nav
      className="navbar navbar-default"
      style={
        Object {
          "borderRadius": 0,
        }
      }>
      <div
        className="container">
        <div
          className="navbar-header">
          <a
            className="navbar-brand"
            href="/"
            style={
              Object {
                "color": "#000",
              }
            }>
            Philosopher Profiles
          </a>
          <button
            className="navbar-toggle collapsed"
            onClick={[Function]}
            type="button">
            <span
              className="sr-only">
              Toggle navigation
            </span>
            <span
              className="icon-bar" />
            <span
              className="icon-bar" />
            <span
              className="icon-bar" />
          </button>
        </div>
        <div
          aria-expanded={null}
          className="navbar-collapse collapse">
          <ul
            className="nav navbar-nav"
            role={null}>
            <li>
              <a
                className=""
                href="#/"
                onClick={[Function]}
                style={Object {}}>
                <span
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#eee",
                      "borderRadius": "4px",
                      "padding": "8px",
                    }
                  }>
                  <span
                    className="glyphicon glyphicon-th" />
                   
                  Grid View
                </span>
              </a>
            </li>
            <li>
              <a
                className=""
                href="#/"
                onClick={[Function]}
                style={Object {}}>
                <span
                  onClick={[Function]}
                  style={
                    Object {
                      "padding": "8px",
                    }
                  }>
                  <span
                    className="glyphicon glyphicon-th-list" />
                   
                  List View
                </span>
              </a>
            </li>
            <li
              className="dropdown"
              style={undefined}>
              <a
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle"
                href="#"
                id="basic-nav-dropdown"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button">
                Filter By
                 
                <span
                  className="caret" />
              </a>
              <ul
                aria-labelledby="basic-nav-dropdown"
                className="dropdown-menu"
                role="menu">
                <div
                  active={undefined}
                  onKeyDown={[Function]}
                  onSelect={[Function]}
                  style={
                    Object {
                      "minWidth": "300px",
                      "padding": "10px",
                    }
                  }>
                  <label
                    className="control-label"
                    htmlFor={undefined}>
                    By Interest
                  </label>
                  <div
                    className="form-group">
                    <div
                      className="checkbox"
                      style={undefined}>
                      <label>
                        <input
                          checked={true}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox" />
                        All
                      </label>
                    </div>
                    <div
                      className="checkbox"
                      style={undefined}>
                      <label>
                        <input
                          checked={false}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox" />
                        Ethics
                      </label>
                    </div>
                    <div
                      className="checkbox"
                      style={undefined}>
                      <label>
                        <input
                          checked={false}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox" />
                        Epistemology
                      </label>
                    </div>
                  </div>
                  <label
                    className="control-label"
                    htmlFor={undefined}>
                    By Category
                  </label>
                  <div
                    className="form-group">
                    <div
                      className="checkbox"
                      style={undefined}>
                      <label>
                        <input
                          checked={false}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox" />
                        Female
                      </label>
                    </div>
                    <div
                      className="checkbox"
                      style={undefined}>
                      <label>
                        <input
                          checked={false}
                          disabled={false}
                          onChange={[Function]}
                          type="checkbox" />
                        Person of Color
                      </label>
                    </div>
                  </div>
                </div>
              </ul>
            </li>
            <li
              className="dropdown"
              style={undefined}>
              <a
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle"
                href="#"
                id="basic-nav-dropdown"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button">
                Sort By
                 
                <span
                  className="caret" />
              </a>
              <ul
                aria-labelledby="basic-nav-dropdown"
                className="dropdown-menu"
                role="menu">
                <li
                  className=""
                  role="presentation"
                  style={undefined}>
                  <a
                    href="#"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1">
                    Last Name
                  </a>
                </li>
                <li
                  className=""
                  role="presentation"
                  style={undefined}>
                  <a
                    href="#"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex="-1">
                    Year
                  </a>
                </li>
              </ul>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </div>
  <div>
    
    <div
      className="container">
      <div
        className="row">
        <div
          className="col-lg-2 col-md-3 col-sm-4 col-xs-6"
          style={
            Object {
              "padding": "5px",
            }
          }>
          <a
            className=""
            href="#/AdaLovelace"
            onClick={[Function]}
            style={
              Object {
                "textDecoration": "none",
              }
            }>
            <div
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#111",
                  "backgroundImage": "url(\"imgs/ada-lovelace.png\")",
                  "backgroundPosition": "center center",
                  "backgroundRepeat": "no-repeat",
                  "backgroundSize": "contain",
                  "borderRadius": "5px",
                  "cursor": "pointer",
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "height": "300px",
                  "overflow": "hidden",
                }
              }>
              <div
                style={
                  Object {
                    "backgroundColor": "#fff",
                    "border": "2px solid #000",
                    "borderRadius": "4px",
                    "color": "#000",
                    "fontFamily": "Avenir",
                    "fontSize": "18px",
                    "fontWeight": "700",
                    "marginBottom": "10px",
                    "padding": "2px 6px",
                    "textAlign": "center",
                    "width": "80%",
                  }
                }>
                <span>
                  Ada
                   
                  Lovelace
                </span>
              </div>
            </div>
          </a>
        </div>
        <div
          className="col-lg-2 col-md-3 col-sm-4 col-xs-6"
          style={
            Object {
              "padding": "5px",
            }
          }>
          <a
            className=""
            href="#/EthicsSocrates"
            onClick={[Function]}
            style={
              Object {
                "textDecoration": "none",
              }
            }>
            <div
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#111",
                  "backgroundImage": "url(\"imgs/socrates.png\")",
                  "backgroundPosition": "center center",
                  "backgroundRepeat": "no-repeat",
                  "backgroundSize": "contain",
                  "borderRadius": "5px",
                  "cursor": "pointer",
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "height": "300px",
                  "overflow": "hidden",
                }
              }>
              <div
                style={
                  Object {
                    "backgroundColor": "#fff",
                    "border": "2px solid #000",
                    "borderRadius": "4px",
                    "color": "#000",
                    "fontFamily": "Avenir",
                    "fontSize": "18px",
                    "fontWeight": "700",
                    "marginBottom": "10px",
                    "padding": "2px 6px",
                    "textAlign": "center",
                    "width": "80%",
                  }
                }>
                <span>
                  Ethics
                   
                  Socrates
                </span>
              </div>
            </div>
          </a>
        </div>
        <div
          className="col-lg-2 col-md-3 col-sm-4 col-xs-6"
          style={
            Object {
              "padding": "5px",
            }
          }>
          <a
            className=""
            href="#/EpisSocrates"
            onClick={[Function]}
            style={
              Object {
                "textDecoration": "none",
              }
            }>
            <div
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#111",
                  "backgroundImage": "url(\"imgs/socrates.png\")",
                  "backgroundPosition": "center center",
                  "backgroundRepeat": "no-repeat",
                  "backgroundSize": "contain",
                  "borderRadius": "5px",
                  "cursor": "pointer",
                  "display": "flex",
                  "flexDirection": "column-reverse",
                  "height": "300px",
                  "overflow": "hidden",
                }
              }>
              <div
                style={
                  Object {
                    "backgroundColor": "#fff",
                    "border": "2px solid #000",
                    "borderRadius": "4px",
                    "color": "#000",
                    "fontFamily": "Avenir",
                    "fontSize": "18px",
                    "fontWeight": "700",
                    "marginBottom": "10px",
                    "padding": "2px 6px",
                    "textAlign": "center",
                    "width": "80%",
                  }
                }>
                <span>
                  Epis
                   
                  Socrates
                </span>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
